lua/nvim-tree/node/init.lua

```diff
@@ -0,0 +1,347 @@
+---@param self BaseNode
+---@param T BaseNode
+---@return boolean
+function BaseNode:is(T)
+  local mt = getmetatable(self)
+  while mt do
+    if mt == T then
+      return true
+    end
+    mt = getmetatable(mt)
+  end
+  return false
+end
+
+---@return boolean
+function BaseNode:has_one_child_folder()
+  return #self.nodes == 1 and self.nodes[1].nodes and vim.loop.fs_access(self.nodes[1].absolute_path, "R") or false
+end
+
+---@param parent_ignored boolean
+---@param status table|nil
+function BaseNode:update_git_status(parent_ignored, status)
+  local get_status
+  if self.nodes then
+    get_status = git.git_status_dir
+  else
+    get_status = git.git_status_file
+  end
+
+  -- status of the node's absolute path
+  self.git_status = get_status(parent_ignored, status, self.absolute_path)
+
+  -- status of the link target, if the link itself is not dirty
+  if self.link_to and not self.git_status then
+    self.git_status = get_status(parent_ignored, status, self.link_to)
+  end
+end
+
+---@return GitStatus|nil
+function BaseNode:get_git_status()
+  if not self.git_status then
+    -- status doesn't exist
+    return nil
+  end
+
+  if not self.nodes then
+    -- file
+    return self.git_status.file and { self.git_status.file }
+  end
+
+  -- dir
+  if not self.explorer.opts.git.show_on_dirs then
+    return nil
+  end
+
+  local status = {}
+  if not self:last_group_node().open or self.explorer.opts.git.show_on_open_dirs then
+    -- dir is closed or we should show on open_dirs
+    if self.git_status.file ~= nil then
+      table.insert(status, self.git_status.file)
+    end
+    if self.git_status.dir ~= nil then
+      if self.git_status.dir.direct ~= nil then
+        for _, s in pairs(self.git_status.dir.direct) do
+          table.insert(status, s)
+        end
+      end
+      if self.git_status.dir.indirect ~= nil then
+        for _, s in pairs(self.git_status.dir.indirect) do
+          table.insert(status, s)
+        end
+      end
+    end
+  else
+    -- dir is open and we shouldn't show on open_dirs
+    if self.git_status.file ~= nil then
+      table.insert(status, self.git_status.file)
+    end
+    if self.git_status.dir ~= nil and self.git_status.dir.direct ~= nil then
+      local deleted = {
+        [" D"] = true,
+        ["D "] = true,
+        ["RD"] = true,
+        ["DD"] = true,
+      }
+      for _, s in pairs(self.git_status.dir.direct) do
+        if deleted[s] then
+          table.insert(status, s)
+        end
+      end
+    end
+  end
+  if #status == 0 then
+    return nil
+  else
+    return status
+  end
+end
+
+---@param projects table
+function BaseNode:reload_node_status(projects)
+  local toplevel = git.get_toplevel(self.absolute_path)
+  local status = projects[toplevel] or {}
+  for _, node in ipairs(self.nodes) do
+    node:update_git_status(self:is_git_ignored(), status)
+    if node.nodes and #node.nodes > 0 then
+      node:reload_node_status(projects)
```

Comment:
Was previously `self:reload_node_status(projects)` resulting in infinite recursion.